Fix version detection

The Metrics fix for #560 unfortunately doesn't work for DSE builds:

```
ERROR [DSE main thread] 2024-10-28 18:34:25,661  MetricsInterceptor.java:92 - Unable to start metrics endpoint
java.lang.NoClassDefFoundError: org/apache/cassandra/utils/CassandraVersion
	at io.k8ssandra.metrics.builder.CassandraMetricRegistryListener.<init>(CassandraMetricRegistryListener.java:89)
	at io.k8ssandra.metrics.prometheus.CassandraDropwizardExports.<init>(CassandraDropwizardExports.java:42)
	at io.k8ssandra.metrics.interceptors.MetricsInterceptor.intercept(MetricsInterceptor.java:71)
	at org.apache.cassandra.service.CassandraDaemon.start(CassandraDaemon.java)
	at com.datastax.bdp.server.DseDaemon.start(DseDaemon.java:559)
	at org.apache.cassandra.service.CassandraDaemon.activate0(CassandraDaemon.java:830)
	at org.apache.cassandra.service.CassandraDaemon.access$100(CassandraDaemon.java:93)
	at org.apache.cassandra.service.CassandraDaemon$3.run(CassandraDaemon.java:738)
Caused by: java.lang.ClassNotFoundException: org.apache.cassandra.utils.CassandraVersion
	at java.net.URLClassLoader.findClass(URLClassLoader.java:387)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:352)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
	... 8 common frames omitted
```

┆Issue is synchronized with this [Jira Story](https://datastax.jira.com/browse/MAPI-78) by [Unito](https://www.unito.io)
┆Issue Number: MAPI-78

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -17,6 +17,7 @@ Changelog for Management API, new PRs should update the `main / unreleased` sect
 * [ENHANCEMENT] [#552](https://github.com/k8ssandra/management-api-for-apache-cassandra/issues/552) Improve "liveness" probe implementation
 * [BUGFIX] [#553](https://github.com/k8ssandra/management-api-for-apache-cassandra/issues/553) Fix CassandraTaskExports metric filtering to make it work with 5.0.x Major compactions
 * [BUGFIX] [#560](https://github.com/k8ssandra/management-api-for-apache-cassandra/issues/560) Fix LatencyMetrics bucketing on 4.1 and 5.0 as their reservoir stores the data in microseconds, not nano (unlike 3.11 and 4.0)
+* [BUGFIX] [#562](https://github.com/k8ssandra/management-api-for-apache-cassandra/issues/562) Fix version detection
 
 ## v0.1.87 (2024-10-02)
 * [FEATURE] [#535](https://github.com/k8ssandra/management-api-for-apache-cassandra/issues/535) Add Cassandra 5.0.0 to the build matrix

--- a/management-api-agent-common/src/main/java/io/k8ssandra/metrics/builder/CassandraMetricRegistryListener.java
+++ b/management-api-agent-common/src/main/java/io/k8ssandra/metrics/builder/CassandraMetricRegistryListener.java
@@ -31,7 +31,6 @@ import java.util.concurrent.ConcurrentHashMap;
 import java.util.function.Supplier;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
-import org.apache.cassandra.utils.CassandraVersion;
 import org.apache.cassandra.utils.EstimatedHistogram;
 import org.apache.cassandra.utils.FBUtilities;
 import org.slf4j.LoggerFactory;
@@ -54,6 +53,8 @@ public class CassandraMetricRegistryListener implements MetricRegistryListener {
       Pattern.compile("([1-9]\\d*)\\.(\\d+)\\.(\\d+)(?:-([a-zA-Z0-9]+))?");
 
   private static final String SERVER_VERSION = FBUtilities.getReleaseVersionString();
+  private static final int SERVER_MAJOR_VERSION;
+  private static final int SERVER_MINOR_VERSION;
   private static final int SERVER_PATCH_VERSION;
 
   private boolean microLatencyBuckets = false;
@@ -61,10 +62,14 @@ public class CassandraMetricRegistryListener implements MetricRegistryListener {
   static {
     Matcher matcher = VERSION_PATTERN.matcher(SERVER_VERSION);
     if (matcher.matches()) {
+      SERVER_MAJOR_VERSION = Integer.parseInt(matcher.group(1));
+      SERVER_MINOR_VERSION = Integer.parseInt(matcher.group(2));
       SERVER_PATCH_VERSION = Integer.parseInt(matcher.group(3));
     } else {
       // unexpected Server version
       logger.warn("Unexpected Server Version string: " + SERVER_VERSION);
+      SERVER_MAJOR_VERSION = -1;
+      SERVER_MINOR_VERSION = -1;
       SERVER_PATCH_VERSION = -1;
     }
   }
@@ -86,12 +91,8 @@ public class CassandraMetricRegistryListener implements MetricRegistryListener {
     parser = CassandraMetricNameParser.getDefaultParser(config);
     cache = new ConcurrentHashMap<>();
 
-    CassandraVersion cassandraVersion = new CassandraVersion(FBUtilities.getReleaseVersionString());
-
-    if (cassandraVersion.major > 4 || (cassandraVersion.major == 4 && cassandraVersion.minor > 0)) {
-      // 4.1 and up should use microsecond buckets
-      microLatencyBuckets = true;
-    }
+    // 4.1 and up should use microsecond buckets
+    microLatencyBuckets = isMicrosecondLatencyBuckets();
 
     this.familyCache = familyCache;
   }
@@ -516,5 +517,23 @@ public class CassandraMetricRegistryListener implements MetricRegistryListener {
   @Override
   public void onTimerRemoved(String name) {
     onHistogramRemoved(name);
+  }
+
+  /**
+   * Returns true if the server version should use microsecond latency buckets, as opposed to
+   * nanosecond latency buckets. For Cassandra 4.1 and newer, this should return true. For Cassandra
+   * 4.0 and older, and DSE 6.8/6.9, this should return false.
+   *
+   * @return true if microsecond latency buckets should be used, false if nanosecond should be used.
+   */
+  private boolean isMicrosecondLatencyBuckets() {
+
+    // This should only catch Cassandra 4.1+ and 5+ versions. DSE 6.8 reports something like
+    // 4.0.0.6851 and 6.9 reports something like 4.0.0.693, both of which would follow C* 4.0.x
+    // nanosecond buckets
+    if (SERVER_MAJOR_VERSION > 4 || (SERVER_MAJOR_VERSION == 4 && SERVER_MINOR_VERSION > 0)) {
+      return true;
+    }
+    return false;
   }
 }

